server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java

```diff
@@ -28,30 +31,32 @@
 @Slf4j
 public class UpstreamRoute {
 
-    private static final long DEFAULT_RETRY_AFTER_SECONDS_VALUE = 30;
-
-    private final LoadBalancer<UpstreamState> balancer;
+    private final TieredBalancer balancer;
     /**
      * The maximum number of attempts the route may retry
      */
     private final int maxRetryAttempts;
 
     /**
-     * Current upstream state
+     * Current upstream
      */
     @Nullable
-    private UpstreamState upstreamState;
+    private Upstream upstream;
     /**
      * Attempt counter
      */
     @Getter
     private int attemptCount;
 
-    public UpstreamRoute(LoadBalancer<UpstreamState> balancer, int maxRetryAttempts) {
+    private final Predicate<Upstream> fallbackPredicate;
+
+    public UpstreamRoute(TieredBalancer balancer, int maxRetryAttempts) {
         this.balancer = balancer;
         this.maxRetryAttempts = maxRetryAttempts;
-        this.upstreamState = balancer.next();
-        this.attemptCount = upstreamState == null ? 0 : 1;
+        Set<Upstream> usedUpstreams = new HashSet<>();
```

Comment:
I do not get it. I expected that you would update it when next() is called. But instead, it is updated in predicate.